File: src/main/java/com/searchlab/ingest/Chunker.java
package com.searchlab.ingest;

import com.knuddels.jtokkit.Encodings;
import com.knuddels.jtokkit.api.Encoding;
import com.knuddels.jtokkit.api.EncodingRegistry;
import com.knuddels.jtokkit.api.EncodingType;
import com.knuddels.jtokkit.api.IntArrayList;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.List;

public class Chunker {

    private static final Logger log = LoggerFactory.getLogger(Chunker.class);
    private static final int CHUNK_SIZE = 512;

    private final Encoding encoding;

    public Chunker() {
        EncodingRegistry registry = Encodings.newDefaultEncodingRegistry();
        this.encoding = registry.getEncoding(EncodingType.CL100K_BASE);
    }

    /**
     * Splits page text into 512-token chunks (no overlap).
     * Each chunk records the page number of its first token.
     */
    public List<Chunk> chunk(List<PageText> pages) {
        // Build a flat token stream, tracking page attribution per token
        List<Integer> tokens = new ArrayList<>();
        List<Integer> tokenPages = new ArrayList<>();

        for (PageText page : pages) {
            IntArrayList pageTokens = encoding.encode(page.text());
            for (int i = 0; i < pageTokens.size(); i++) {
                tokens.add(pageTokens.get(i));
                tokenPages.add(page.pageNumber());
            }
        }

        List<Chunk> chunks = new ArrayList<>();
        int totalTokens = tokens.size();
        int position = 0;

        for (int start = 0; start < totalTokens; start += CHUNK_SIZE) {
            int end = Math.min(start + CHUNK_SIZE, totalTokens);
            IntArrayList window = new IntArrayList(end - start);
            for (int i = start; i < end; i++) {
                window.add(tokens.get(i));
            }
            String text = encoding.decode(window);
            int pageNumber = tokenPages.get(start);
            chunks.add(new Chunk(text, pageNumber, position++));
        }

        log.debug("Chunked {} tokens into {} chunks", totalTokens, chunks.size());
        return chunks;
    }
}

File: service/searchlab/ingest/chunker.py
from dataclasses import dataclass

import tiktoken

from .pdf_parser import PageText

_ENCODING = None
CHUNK_SIZE = 512


def _enc():
    global _ENCODING
    if _ENCODING is None:
        _ENCODING = tiktoken.get_encoding("cl100k_base")
    return _ENCODING


@dataclass
class Chunk:
    text: str
    page_number: int
    position: int


def chunk(pages: list[PageText]) -> list[Chunk]:
    enc = _enc()

    tokens: list[int] = []
    token_pages: list[int] = []

    for page in pages:
        page_tokens = enc.encode(page.text)
        tokens.extend(page_tokens)
        token_pages.extend([page.page_number] * len(page_tokens))

    chunks: list[Chunk] = []
    position = 0
    for start in range(0, len(tokens), CHUNK_SIZE):
        window = tokens[start : start + CHUNK_SIZE]
        text = enc.decode(window)
        page_num = token_pages[start]
        chunks.append(Chunk(text=text, page_number=page_num, position=position))
        position += 1

    return chunks
